### ISSUE TITLE
Feature Request: 8x6 Tech Inventory for Ships and Microsoft Store Save support

### ISSUE THREAD

**Issue by SirCathal** [goatfungus/NMSSaveEditor#444]
Hi there,
I have also tested another NMS editor (NomNom). It is less comfortable to use and significantly slower than this Editor, but it has two features that I sorely miss here.
1) You can increase the technology inventory of the ships up to 8x6, here it only goes up to 7x4. If you have done that with it, the technology inventory is also displayed correctly in NMSSaveEditor with 8x6 and is fully usable.
2) NomNom also supports the Save Format of the Microsoft Store version of NMS. That has a very big advantage. If you play NMS Digital on the XBox you can play on Windows using the Microsoft Play Anywhere feature. The saves are synchronized. That means if I edit the game status on the PC I can also play it on the console and have one of the few possibilities to edit savegames on the console. 

@goatfungus Do you think you can implement those two features?

The Windows Store Saves are now located here:
c:\Users\<WindwosUserName>\AppData\Local\Packages\HelloGames.NoMansSky_bs190hzg1sesy\SystemAppData\wgs\

**Comment by goatfungus** [comment]
Can you send me a copy of your saves so I can compare structure?

**Comment by SirCathal** [comment]
Yes, I posted them in this issue:
https://github.com/goatfungus/NMSSaveEditor/issues/419#issuecomment-917607512

**Comment by goatfungus** [comment]
I have been working on Windows Store support, and I was wondering if you're still up for testing what I've done?

If so, you can download a demo here:
[NMSSaveEditor.zip](https://github.com/goatfungus/NMSSaveEditor/files/7169542/NMSSaveEditor.zip)

NOTE: Please backup all your files before attempting to use this version, as it includes quite a large overhaul of the storage system which may corrupt your files.

**Comment by SirCathal** [comment]
> I have been working on Windows Store support, and I was wondering if you're still up for testing what I've done?
> 
> If so, you can download a demo here:
> [NMSSaveEditor.zip](https://github.com/goatfungus/NMSSaveEditor/files/7169542/NMSSaveEditor.zip)
@goatfungus Thanx, but I think you uploaded the wrong file, that Version inside the ZIP is very old (1.5) and much smaller then the current Version. It can't find the Windows Store Saves when I select the folder with the containers.index File.

**Comment by goatfungus** [comment]
Sorry, try this one:
[NMSSaveEditor.zip](https://github.com/goatfungus/NMSSaveEditor/files/7172897/NMSSaveEditor.zip)

**Comment by Voidw4rd** [comment]
Right now I'm doing tests. I have the NMS version of Game Pass and it has already identified the salvage files (I suppose that is an advance), now I am verifying that everything is fine. And load the game later.

PS: Sorry for my bad English is not my native language

**Comment by SirCathal** [comment]
Yes this version works. Here what I have tested and what other minor problems I had.
Editing of what has been learned, spaceships, Exo Suite, Muiltitool and the general values ​​works well.
Importing a complete spaceship from a GOG or Steam saved game also works great.

I had a little problem when I restarted the game and created a new save (I tested it with a new creative save and a second Expedition save).

When I start the editor afterwards, I get a warning in the shell output, which says that exactly the newly created savegame could not be read, the first time also an error message shows up:
```
Initializing environment ...
Starting Editor ...
Reading Container Index
  AccountData
  Settings
  Slot1Auto
  Slot1Manual
  Slot2Auto
  Slot2Manual
  Slot3Auto
AccountData
WARNING: Cannot read account data
Slot1Auto
  version: 1
  totalPlayTime: 172: 17
Slot1Manual
  version: 1
  totalPlayTime: 170: 51
Slot2Auto
  version: 196614
  totalPlayTime: 10:20
Slot2Manual
  version: 196614
  totalPlayTime: 10:55
Slot3Auto
WARNING: cannot read container: Slot3Auto
java.lang.NullPointerException
        at nomanssave.Application. <init> (Unknown Source)
        at nomanssave.Application. <init> (Unknown Source)
        at nomanssave.G.run (Unknown Source)
        at java.awt.event.InvocationEvent.dispatch (Unknown Source)
        at java.awt.EventQueue.dispatchEventImpl (Unknown Source)
        at java.awt.EventQueue.access $ 500 (Unknown Source)
        at java.awt.EventQueue $ 3.run (Unknown Source)
        at java.awt.EventQueue $ 3.run (Unknown Source)
        at java.security.AccessController.doPrivileged (Native Method)
        at java.security.ProtectionDomain $ JavaSecurityAccessImpl.doIntersectionPrivilege (Unknown Source)
        at java.awt.EventQueue.dispatchEvent (Unknown Source)
        at java.awt.EventDispatchThread.pumpOneEventForFilters (Unknown Source)
        at java.awt.EventDispatchThread.pumpEventsForFilter (Unknown Source)
        at java.awt.EventDispatchThread.pumpEventsForHierarchy (Unknown Source)
        at java.awt.EventDispatchThread.pumpEvents (Unknown Source)
        at java.awt.EventDispatchThread.pumpEvents (Unknown Source)
        at java.awt.EventDispatchThread.run (Unknown Source)
Press any key . . .

```
The previously mentioned save slot is then also not offered for selection.

But after that, if I just quit the editor and start it again, there are no more warnings, the new savegame is displayed and can also be edited without errors later in the game.

If I then create another new savegame, a warning comes up again.

```
Initializing environment ...
Starting Editor ...
Reading Container Index
  AccountData
  Settings
  Slot1Auto
  Slot1Manual
  Slot2Auto
  Slot2Manual
  Slot3Auto
  Slot4Auto
AccountData
Slot1Auto
  version: 1
  totalPlayTime: 172: 17
Slot1Manual
  version: 1
  totalPlayTime: 170: 51
Slot2Auto
  version: 196614
  totalPlayTime: 10:20
Slot2Manual
  version: 196614
  totalPlayTime: 10:55
Slot3Auto
  version: 2
  totalPlayTime: 0:37
Slot4Auto
WARNING: cannot read container: Slot4Auto
  using name mappings: NMS 3.6
```

After quitting and restarting the editor, the warning has disappeared and the savegame can be edited.
In other Words those Warnings only shows up only once for every new created savegame.

I noticed another error while editing, but it is also included in version 1.8.7.
If I insert something into a previously empty warehouse, suit or spaceship slot (e.g. oxygen) then it will only be filled with 250/250 units and not with 9999/9999 as before. It is also not possible to increase the maximum size of the slot for this. In the game itself you can add more, there 9999 is used as the maximum size.

Even if you probably haven't changed it yet. The maximum size of the spacecraft technology inventory is still 7x4 and not 8x6 as it could be.

Overall I have to say great work @goatfungus

**Comment by goatfungus** [comment]
I'll deal with the other issues separately, first I want to make sure all this works properly.
Can you send me a copy of the log file itself?

**Comment by goatfungus** [comment]
Updated version with a few other fixes:
[NMSSaveEditor.zip](https://github.com/goatfungus/NMSSaveEditor/files/7174726/NMSSaveEditor.zip)

I have tweaked the logging part so that any "cannot read container" messages will spit out a trace, and fixed that NullPointerException. This version should also support backup/restore for these save files too.

**Comment by Voidw4rd** [comment]
A question. Today is the first day that I have used the Save Editor. When I edit my freighter, is there a way to add a specific technology? Because if there is in this version it is not appearing.

**Comment by goatfungus** [comment]
> A question. Today is the first day that I have used the Save Editor. When I edit my freighter, is there a way to add a specific technology? Because if there is in this version it is not appearing.

Usually just right-click on a technology slot and "Add Item".

![image](https://user-images.githubusercontent.com/41274031/133546617-7e3d71f2-fed9-4266-b9d8-6a8e3f0f1d81.png)

**Comment by Voidw4rd** [comment]
![image](https://user-images.githubusercontent.com/67162397/133546927-d0d8475f-b6b6-450d-b117-381a2cccdb18.png)


Yes, it was what I feared in this version it does not appear.

**Comment by Voidw4rd** [comment]
If it was what I was afraid of, in this version the technology segment does not appear

![image](https://user-images.githubusercontent.com/67162397/133547043-6d0023dc-ce8a-4cc7-9326-8d22974a2675.png)

**Comment by goatfungus** [comment]
Have you tried the "Technology" section?

![image](https://user-images.githubusercontent.com/41274031/133548676-c8db58fc-b04a-488d-8c53-2cba3be71621.png)

**Comment by goatfungus** [comment]
In any case, I think Technology should be available on the General area as well.... leave it with me and I'll see what I can do.

**Comment by Voidw4rd** [comment]
Yep. I tried on tecnology section, but nothing. I use the search box and didn't worrk :P

**Comment by goatfungus** [comment]
Will add that to next update along with the MS Game Pass stuff, but that is priority at the moment.

**Comment by goatfungus** [comment]
For those of you who want to test MS Game Pass saves, try this experimental build:
[NMSSaveEditor.zip](https://github.com/goatfungus/NMSSaveEditor/files/7174977/NMSSaveEditor.zip)

I am going away for the weekend, so I won't be working on any more changes until I get back, but I think it is mostly stable now. Please leave a message below if you encounter any issues, and be sure to attach your NMSSaveEditor.log file and any other information I might need to help solve any issues.

**Comment by Voidw4rd** [comment]
![image](https://user-images.githubusercontent.com/67162397/133551297-14229463-f4c4-4a3e-a70a-95d3e215bfcc.png)

I was editing fine with no problems. But from one moment to the next the bran stopped working. The advantage is that it does not corrupt any of the files, it just seems that it does not have access to them.
[NMSSaveEditor.log](https://github.com/goatfungus/NMSSaveEditor/files/7174992/NMSSaveEditor.log)

Have a good weekend =) TY for the effort if I could help just say so...

**Comment by goatfungus** [comment]
Can you try creating a folder called "backups" in the same location as where you put the editor files?

The error suggests the backup folder does not exist, and it's not trying to create it for whatever reason.

**Comment by Santalan99** [comment]
Hi! i've been trying to use the editor for WMS NMS and I can open the save file edit but i can not save any changes i get "There was an error saving the file" I'm I doing anything wrong? even if I don't make any changes to the save I still get the error :S

**Comment by goatfungus** [comment]
If you're using this experimental build, create a "backups" folder. I will fix this before it goes live, but for now this is an easy workaround.

**Comment by goatfungus** [comment]
> Hi! i've been trying to use the editor for WMS NMS and I can open the save file edit but i can not save any changes i get "There was an error saving the file" I'm I doing anything wrong? even if I don't make any changes to the save I still get the error :S

If you're using this experimental build, create a "backups" folder. I will fix this before it goes live, but for now this is an easy workaround.

**Comment by Voidw4rd** [comment]
Yeah. Fixed. Sorry for the late answer I was editing a save. I read the log and see that. =) have a great weekend TY for all

**Comment by Santalan99** [comment]
Yeah i have a backup of my files, if you need a guy that "monkey tests" the experimental build I'm your guy

**Comment by Santalan99** [comment]
> > Hi! i've been trying to use the editor for WMS NMS and I can open the save file edit but i can not save any changes i get "There was an error saving the file" I'm I doing anything wrong? even if I don't make any changes to the save I still get the error :S
> 
> If you're using this experimental build, create a "backups" folder. I will fix this before it goes live, but for now this is an easy workaround.

It worked like a charm :)

**Comment by SirCathal** [comment]
The latest Version works good, the first start problem with new saves is gone.

The only things I found are a Warning after starting the Editor:
```
  Slot: Slot 5 - NORMAL - Sep 16, 01:15
  Filename: Slot5Auto
  using name mappings: NMS 3.6
  WARNING: name mapping not found: D=h
```

But during the test, sometimes, after saving in the Game, I got the message 
"cannot read container: Slot5Manual" (where Slot number is the latest used save)

```
[INFO] Slot5Manual
[DEBUG] header: 4
[DEBUG] name: data
[DEBUG] filename: 5ABCA1FBE6774C0794A93C14F79133EC
[DEBUG] filename2: B385227F38E34F90AD6FBDA843467CB7
[DEBUG] name: meta
[DEBUG] filename: B3E65764FC424529A0AD5385EC4AFEC7
[DEBUG] filename2: 4064E0EFF9B34BE29FB7E9C49422348A
[WARNING] cannot read container: Slot5Manual
[WARNING] java.io.IOException: data size mismatch
	at nomanssave.storage.w.<init>(Unknown Source)
	at nomanssave.storage.x.<init>(Unknown Source)
	at nomanssave.storage.y.<init>(Unknown Source)
	at nomanssave.storage.s.<init>(Unknown Source)
	at nomanssave.storage.g.a(Unknown Source)
	at nomanssave.Application.<init>(Unknown Source)
	at nomanssave.Application.<init>(Unknown Source)
	at nomanssave.G.run(Unknown Source)
	at java.awt.event.InvocationEvent.dispatch(Unknown Source)
	at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
	at java.awt.EventQueue.access$500(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue.dispatchEvent(Unknown Source)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.run(Unknown Source)
[DEBUG] Save Path: c:\Users\naclador\AppData\Local\Packages\HelloGames.NoMansSky_bs190hzg1sesy\SystemAppData\wgs\00090000074A00FC_29070100B936489ABCE8B9AF3980429C
[DEBUG] Mem Usage: 16/260 MB
[DEBUG] Latest version: "1.8.7"
[DEBUG] Current version: "1.9.0"

```
[NMSSaveEditor.log](https://github.com/goatfungus/NMSSaveEditor/files/7179479/NMSSaveEditor.log)

And again, when I close the Editor and start it again, it work, because the editor preselect a backup Save File. But I can then switch to Slot5Auto or Slot5Manual and everything works fine.

**Comment by goatfungus** [comment]
So if i'm getting this correct, if you leave the editor open while you're playing the game and the game does a save the editor hits that error?
This will be because the editor does not support automatic reloading for MS Game Pass files yet.

**Comment by SirCathal** [comment]
No, I always close the Editor before I play and I always close the Game before I edit.

This happens sometimes after closing the Game and then starting the Editor.

Maybe that comes together with another issue I found.
When I save the Game with the Editor, It will be saved in the future time. something adds 12 hours to the actual date and Time.
For example, when I edit the Game at 13:00 at 18.09.2021 the Game displays that save as 01:00 19.09.2021.

When I then play the game and save it inside the Game. It saves it with the correct Date.
After that, the editor lists the Backup before the normal Saves and that is a bit confusing and dangerous, because one time I did not see that, edited the backup because it was automaticly selected and saved that old save over my newer one.

**Comment by goatfungus** [comment]
Ah, i think i know what is causing that.... i'll fix it as soon as i can

**Comment by goatfungus** [comment]
Ok, I have released a new version (1.9.1) which should now be able to read and write WGS save files.
Please remember to back up your files before using this though.

**Comment by SirCathal** [comment]
Cool, I will test more today after Work. The Bigger Ship Inventory works fine and the Save Date seems to correct.
I did a lot of testing with the 1.9.0 Version over the weekend and It seems to be working fine. All edits of the windows Store Version are synced to the XBox save automaticly after starting the Windows Store Version, thats great! 
The Only thing (yes it's another issue) that's left over for me is the to small slot size of 250/250 and not 9999/9999 when adding stuff.

**Comment by goatfungus** [comment]
Great to know most of it works now :)
Item sizes are actually taken from the game files themselves, and I believe there is a difference depending on game mode as well. I'll have a look into that when i get a chance.

**Comment by SirCathal** [comment]
Ok, that makes sense, it seems to be a Issue with Expedition saves (on all Plattforms) but then something is wrong defined because ingame 9999/9999 is the limit for Expedition games.

**Comment by goatfungus** [comment]
I have updated the stack size logic so that it now detects the base type of Expedition saves (this can be Normal, Survival, Permadeath, etc). Version 1.9.2 should now work properly for Expedition saves.
